File: src/services/ai.ts
import { invoke } from "@tauri-apps/api/core";

/** Configuration for a single AI role (STT, TTS or chat). */
export interface AiProviderConfig {
  provider: string;
  api_key: string;
  model: string;
  base_url: string;
}

/** AI model provider settings for each role. */
export interface AiSettings {
  // Speech-to-text (transcription)
  stt_provider: string;
  stt_api_key: string;
  stt_model: string;
  stt_base_url: string;

  // Text-to-speech
  tts_provider: string;
  tts_api_key: string;
  tts_model: string;
  tts_base_url: string;

  // Chat / summarization (LLM)
  chat_provider: string;
  chat_api_key: string;
  chat_model: string;
  chat_base_url: string;
}

/**
 * Sensible defaults applied when a field is left blank. These mirror the
 * OpenAI-compatible API shape that most providers (OpenAI, Ollama, LM Studio,
 * etc.) expose, so a user only needs to override what differs for their setup.
 */
export const AI_DEFAULTS: AiSettings = {
  stt_provider: "openai",
  stt_api_key: "",
  stt_model: "whisper-1",
  stt_base_url: "https://api.openai.com/v1",

  tts_provider: "openai",
  tts_api_key: "",
  tts_model: "tts-1",
  tts_base_url: "https://api.openai.com/v1",

  chat_provider: "openai",
  chat_api_key: "",
  chat_model: "gpt-4o-mini",
  chat_base_url: "https://api.openai.com/v1",
};

/** The saved AI provider settings (blank defaults when never configured). */
export const getAiSettings = () => invoke<AiSettings>("get_ai_settings");

/** Persist the user's AI provider settings. */
export const setAiSettings = (settings: AiSettings) =>
  invoke<void>("set_ai_settings", { settings });

/**
 * List the models installed on a local / self-hosted server (Ollama, LM Studio,
 * …) reachable at `baseUrl`. Used to populate the model picker for the "local"
 * provider. Throws if the server is unreachable; callers fall back to free text.
 */
export const listLocalModels = (baseUrl: string) =>
  invoke<string[]>("list_local_models", { baseUrl });

File: src-tauri/src/commands/ai.rs
//! Commands for reading and persisting AI model provider settings.
//!
//! Covers the speech-to-text, text-to-speech and chat/LLM providers, each with
//! its own provider name, API key, model name and base URL. Stored as the
//! single row (`id = 1`) of the `ai_settings` table.

use sqlx::{Pool, Row, Sqlite};

use crate::db::pool;
use crate::error::Result;
use crate::settings::AiSettings;

/// Read the saved AI provider settings from the database, or blank defaults when
/// no row exists yet. Shared by [`get_ai_settings`] and other commands (e.g. the
/// summary generator) that need the configured Chat provider.
pub async fn fetch_ai_settings(pool: &Pool<Sqlite>) -> Result<AiSettings> {
    let row = sqlx::query(
        "SELECT stt_provider, stt_api_key, stt_model, stt_base_url,
                tts_provider, tts_api_key, tts_model, tts_base_url,
                chat_provider, chat_api_key, chat_model, chat_base_url
         FROM ai_settings WHERE id = 1",
    )
    .fetch_optional(pool)
    .await?;

    Ok(match row {
        Some(r) => AiSettings {
            stt_provider: r.get("stt_provider"),
            stt_api_key: r.get("stt_api_key"),
            stt_model: r.get("stt_model"),
            stt_base_url: r.get("stt_base_url"),
            tts_provider: r.get("tts_provider"),
            tts_api_key: r.get("tts_api_key"),
            tts_model: r.get("tts_model"),
            tts_base_url: r.get("tts_base_url"),
            chat_provider: r.get("chat_provider"),
            chat_api_key: r.get("chat_api_key"),
            chat_model: r.get("chat_model"),
            chat_base_url: r.get("chat_base_url"),
        },
        None => AiSettings::default(),
    })
}

/// Return the saved AI provider settings, or blank defaults when none are set.
#[tauri::command]
pub async fn get_ai_settings(app: tauri::AppHandle) -> Result<AiSettings> {
    let pool = pool(&app).await?;
    fetch_ai_settings(&pool).await
}

/// Persist the AI provider settings.
#[tauri::command]
pub async fn set_ai_settings(app: tauri::AppHandle, settings: AiSettings) -> Result<()> {
    let pool = pool(&app).await?;
    sqlx::query(
        "INSERT INTO ai_settings
             (id, stt_provider, stt_api_key, stt_model, stt_base_url,
              tts_provider, tts_api_key, tts_model, tts_base_url,
              chat_provider, chat_api_key, chat_model, chat_base_url)
         VALUES (1, $1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12)
         ON CONFLICT(id) DO UPDATE SET
             stt_provider  = excluded.stt_provider,
             stt_api_key   = excluded.stt_api_key,
             stt_model     = excluded.stt_model,
             stt_base_url  = excluded.stt_base_url,
             tts_provider  = excluded.tts_provider,
             tts_api_key   = excluded.tts_api_key,
             tts_model     = excluded.tts_model,
             tts_base_url  = excluded.tts_base_url,
             chat_provider = excluded.chat_provider,
             chat_api_key  = excluded.chat_api_key,
             chat_model    = excluded.chat_model,
             chat_base_url = excluded.chat_base_url",
    )
    .bind(settings.stt_provider)
    .bind(settings.stt_api_key)
    .bind(settings.stt_model)
    .bind(settings.stt_base_url)
    .bind(settings.tts_provider)
    .bind(settings.tts_api_key)
    .bind(settings.tts_model)
    .bind(settings.tts_base_url)
    .bind(settings.chat_provider)
    .bind(settings.chat_api_key)
    .bind(settings.chat_model)
    .bind(settings.chat_base_url)
    .execute(&pool)
    .await?;
    Ok(())
}
